doc/ipmc/IPMC_Data_Plane_HLD.md

```diff
@@ -0,0 +1,472 @@
+
+mgmanager
+
+- IP Multicast route encapsulation management of incoming and outgoing interface groups
+- Encapsulate and manage the creation, modification, and deletion of rpf group and ipmc group
+- Implement rpf/ipmc group and member reference counting, exception handling, and other behaviors
+- Internally invoke the SAI API of the rpf/ipmc group and member. For details about the SAI API, see the following
+- Provides external query and update interfaces for the rpf/ipmc group
+
+mrouteorch
+
+- Listens to MROUTE_TABLE in the APPL DB and queries the corresponding rpf group and ipmc group
+- Associate the corresponding rpf group and ipmc group IDs, and invoke the SAI API to create IP multicast routes.
+
+#### mgmanager
+
+Follow the nexthop group design, and manage the Group and Members separately:
+> The reason why the nexthop group is not used directly, is that the nhg has been strongly coupled with SAI API when it is implemented. The multicast routing services cannot be reused.
+> In addition, nhg IDs are often referenced by other businesses, so abstracting a single mgmanager can also avoid misuse by unfamiliar businesses
+
+- The MgManager exposes group query and management interfaces
+- Internally, RpfMember and RpfGroup manage reference counting and exception handling, invoke the SAI API to generate the corresponding group id
+
+#### Warm reboot requirements
+
+This document mainly considers the data plane IP multicast routing function data plane warm reboot:
+
+The warm reboot support for the protocol plane is not in the design scope. In this case, the IP multicast route obtained from the kernel can be regarded as the latest routing data on the data plane.
+
+1. The data plane is warm rebooted
+2. mroutesycnd reads the kernel IP multicast routing information and sends it to the common warm reboot component for comparison
+3. The warm reboot component writes the data to be updated to the APPL DB to complete the IP multicast route update
+4. The warm reboot of the data plane is complete
+
+#### Sequence diagram
+
+Add an IP multicast route (simplify the protocol plane into an FRR component for description. In actual situations, the selection of the protocol plane may be different):
+
+![Add new IP multicast routes](images/add_ipmc_entry.png "Figure 1: Add new IP multicast routes")
+
+Figure 3.  Add new IP multicast routes.
+
+Warm Reboot:
+
+![IP multicast routes warm reboot](images/ipmc_warm_reboot2.png "Figure 1: IP multicast routes warm reboot")
+
+Figure 4. IP multicast routes warm reboot.
+
+#### Scalability and performance requirements
+
+In terms of capacity, the SAI API does not implement a query interface for the capacity of IP multicast routes and multicast member groups like ECMP routes. Therefore, the relevant capacity is not restricted. However, the device must continue to run normally in scenarios where the capacity is exceeded.

```

Comment:
SAI switch attribute SAI_SWITCH_ATTR_AVAILABLE_IPMC_ENTRY is available to query capacity of IP multicast routes, please include.

Reply:
> SAI switch attribute SAI_SWITCH_ATTR_AVAILABLE_IPMC_ENTRY is available to query capacity of IP multicast routes, please include.

OK, it will be revised in the next commit. The focus here is on the group size and the size of member within each group